// rclnodejs/generated/pi_gpio_interface/pi_gpio_interface__action__GPIO.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
class pi_gpio_interface__action__GPIO {
  static get Goal() {
    return require('./pi_gpio_interface__action__GPIO_Goal.js');
  }
  static get Result() {
    return require('./pi_gpio_interface__action__GPIO_Result.js');
  }
  static get Feedback() {
    return require('./pi_gpio_interface__action__GPIO_Feedback.js');
  }
  static get impl() {
    return {
      get SendGoalService() {
        return require('./pi_gpio_interface__action__GPIO_SendGoal.js');
      },
      get GetResultService() {
        return require('./pi_gpio_interface__action__GPIO_GetResult.js');
      },
      get FeedbackMessage() {
        return require('./pi_gpio_interface__action__GPIO_FeedbackMessage.js');
      },
      get CancelGoal() {
        return require('../action_msgs/action_msgs__srv__CancelGoal.js');
      },
      get GoalStatusArray() {
        return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
      }
    };
  }
  static type() {
    return {pkgName: 'pi_gpio_interface', subFolder: 'action', interfaceName: 'GPIO'};
  }
}
module.exports = pi_gpio_interface__action__GPIO;
