// rclnodejs/generated/pi_gpio_interface/pi_gpio_interface__action__GPIO_Feedback.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GPIO_FeedbackRefStruct = StructType({
  feedback: primitiveTypes.int32,
});
const GPIO_FeedbackRefArray = ArrayType(GPIO_FeedbackRefStruct);
const GPIO_FeedbackRefStructArray = StructType({
  data: GPIO_FeedbackRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GPIO_FeedbackWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GPIO_FeedbackRefStruct(other._refObject.toObject());
      this._feedbackIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GPIO_FeedbackRefStruct();
    this._feedbackIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GPIO_FeedbackWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GPIO_FeedbackArrayWrapper;
  }
  static get ArrayType() {
    return GPIO_FeedbackArrayWrapper;
  }
  static get refObjectArrayType() {
    return GPIO_FeedbackRefStructArray
  }
  static get refObjectType() {
    return GPIO_FeedbackRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._feedbackIntialized) {
        throw new TypeError('Invalid argument: feedback in GPIO_Feedback');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._feedbackIntialized = true;
    this._refObject.feedback = refObject.feedback;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    GPIO_FeedbackWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'pi_gpio_interface', subFolder: 'action', interfaceName: 'GPIO_Feedback'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get feedback() {
    if (!this._feedbackIntialized) {
      return undefined;
    }
    return this._refObject.feedback;
  }
  set feedback(value) {
    this._feedbackIntialized = true;
    this._refObject.feedback = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GPIO_FeedbackRefStruct(refObject.toObject());
    this._feedbackIntialized = true;
  }
  copy(other) {
    this._refObject = new GPIO_FeedbackRefStruct(other._refObject.toObject());
    this._feedbackIntialized = true;
  }
  static get classType() {
    return GPIO_FeedbackWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"feedback","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"pi_gpio_interface","type":"GPIO_Feedback","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GPIO_Feedback"};
  }
  hasMember(name) {
    let memberNames = ["feedback"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GPIO_FeedbackArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GPIO_FeedbackWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GPIO_FeedbackWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GPIO_FeedbackArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GPIO_FeedbackArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GPIO_FeedbackRefArray(size);
    this._refObject = new GPIO_FeedbackRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GPIO_FeedbackWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GPIO_FeedbackArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GPIO_FeedbackArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GPIO_FeedbackWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GPIO_FeedbackWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GPIO_FeedbackArrayWrapper;
  }
}
module.exports = GPIO_FeedbackWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "feedback",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "pi_gpio_interface",
    "type": "GPIO_Feedback",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GPIO_Feedback"
}
*/
