// rclnodejs/generated/ros2_pi_gpio/ros2_pi_gpio__action__GPIO.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
class ros2_pi_gpio__action__GPIO {
  static get Goal() {
    return require('./ros2_pi_gpio__action__GPIO_Goal.js');
  }
  static get Result() {
    return require('./ros2_pi_gpio__action__GPIO_Result.js');
  }
  static get Feedback() {
    return require('./ros2_pi_gpio__action__GPIO_Feedback.js');
  }
  static get impl() {
    return {
      get SendGoalService() {
        return require('./ros2_pi_gpio__action__GPIO_SendGoal.js');
      },
      get GetResultService() {
        return require('./ros2_pi_gpio__action__GPIO_GetResult.js');
      },
      get FeedbackMessage() {
        return require('./ros2_pi_gpio__action__GPIO_FeedbackMessage.js');
      },
      get CancelGoal() {
        return require('../action_msgs/action_msgs__srv__CancelGoal.js');
      },
      get GoalStatusArray() {
        return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
      }
    };
  }
  static type() {
    return {pkgName: 'ros2_pi_gpio', subFolder: 'action', interfaceName: 'GPIO'};
  }
}
module.exports = ros2_pi_gpio__action__GPIO;
