// rclnodejs/generated/ros2_pi_gpio/ros2_pi_gpio__action__GPIO_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GPIO_GoalRefStruct = StructType({
  gpio: primitiveTypes.string,
});
const GPIO_GoalRefArray = ArrayType(GPIO_GoalRefStruct);
const GPIO_GoalRefStructArray = StructType({
  data: GPIO_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GPIO_GoalWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GPIO_GoalRefStruct(other._refObject.toObject());
      this._gpioIntialized = true;
      this._wrapperFields.gpio =  new StringWrapper(other._wrapperFields.gpio);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GPIO_GoalRefStruct();
    this._gpioIntialized = false;
    this._wrapperFields.gpio = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GPIO_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GPIO_GoalArrayWrapper;
  }
  static get ArrayType() {
    return GPIO_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return GPIO_GoalRefStructArray
  }
  static get refObjectType() {
    return GPIO_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._gpioIntialized) {
        throw new TypeError('Invalid argument: gpio in GPIO_Goal');
      }
    }
    if (own) {
      this._wrapperFields.gpio.freeze(own, checkConsistency);
    }
    this._refObject.gpio = this._wrapperFields.gpio.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._gpioIntialized = true;
    this._wrapperFields.gpio.data = refObject.gpio.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.gpio);
  }
  static destoryRawROS(msg) {
    GPIO_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ros2_pi_gpio', subFolder: 'action', interfaceName: 'GPIO_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get gpio() {
    if (!this._gpioIntialized) {
      return undefined;
    }
    return this._wrapperFields.gpio.data;
  }
  set gpio(value) {
    this._gpioIntialized = true;
    this._wrapperFields.gpio.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GPIO_GoalRefStruct(refObject.toObject());
    this._gpioIntialized = true;
    this._wrapperFields.gpio.copyRefObject(this._refObject.gpio);
  }
  copy(other) {
    this._refObject = new GPIO_GoalRefStruct(other._refObject.toObject());
    this._gpioIntialized = true;
    this._wrapperFields.gpio.copy(other._wrapperFields.gpio);
  }
  static get classType() {
    return GPIO_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"gpio","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"ros2_pi_gpio","type":"GPIO_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GPIO_Goal"};
  }
  hasMember(name) {
    let memberNames = ["gpio"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GPIO_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GPIO_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GPIO_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GPIO_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GPIO_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GPIO_GoalRefArray(size);
    this._refObject = new GPIO_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GPIO_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GPIO_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GPIO_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GPIO_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GPIO_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GPIO_GoalArrayWrapper;
  }
}
module.exports = GPIO_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "gpio",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "ros2_pi_gpio",
    "type": "GPIO_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GPIO_Goal"
}
*/
